### PR TITLE
Automatically fills import statement and display autocomplete popup

### PR DESCRIPTION
**Pull request by aliev**
Sorry for another one PR

### CODE REVIEW

**Comment by davidhalter** [comment]
I think that looks good. Agreed @blueyed ?

File: doc/jedi-vim.txt [hunk 1]
```diff
@@ -481,6 +482,18 @@ Function: `jedi#force_py_version(py_version)`
 
 Options: 2 or 3
 Default: "auto" (will use sys.version_info from "python" in your $PATH)
+------------------------------------------------------------------------------
+6.12. `g:jedi#smart_auto_mappings`	            *g:jedi#smart_auto_mappings*
+
+When you start typing from module.name<space> jedi automatically
```
Comment (blueyed) [inline doc/jedi-vim.txt]: s/jedi/jedi-vim/

And put the example into quotes (or backticks?), please.

File: doc/jedi-vim.txt [hunk 2]
```diff
@@ -481,6 +482,18 @@ Function: `jedi#force_py_version(py_version)`
...
+fills import statement and display autocomplete popup.
```
Comment (blueyed) [inline doc/jedi-vim.txt]: s/fills import/adds the "import"/
s/displays/displays the/

File: doc/jedi-vim.txt [hunk 3]
```diff
@@ -481,6 +482,18 @@ Function: `jedi#force_py_version(py_version)`
...
+
+This option can be disabled in the .vimrc:
+
+let g:jedi#smart_auto_mappings = 1
```
Comment (blueyed) [inline doc/jedi-vim.txt]: This should be 0.

**Comment by blueyed** [comment]
Ok with me.

@aliev 
Can you fix the typos/doc and squash your commits?
You can then force-push the branch - there is no need to create a new PR.

Also, if you don't like to do this, I'd be happy to help you and do this when merging it.  Just let me/us know.

Thanks!

**Comment by aliev** [comment]
@blueyed I updated the documentation.

> Also, if you don't like to do this, I'd be happy to help you and do this when merging it. Just let me/us know.

I think you can do it more faster than I. Thanks! :)

**Comment by blueyed** [comment]
Ok.

What do you think about the following change to it?

It will limit the search to max. 50 bytes/chars (for the module name), and only matches directly behind it.

I found it odd that `from foo   _` would complete to `from foo    import _`.

We could normalize/strip the extraneous whitespace, but then again it might be a good workaround if you need to skip it, because it might be slow etc.

    diff --git i/autoload/jedi.vim w/autoload/jedi.vim
    index affc590..03bffd0 100644
    --- i/autoload/jedi.vim
    +++ w/autoload/jedi.vim
    @@ -425,7 +425,7 @@ endfunction
     
     function! jedi#smart_auto_mappings(is_popup_on_dot)
         " Auto put import statement after from module.name<space> and complete
    -    if search('\<from\s\+[A-Za-z0-9._]\+\s*\%#\s*$', 'bcn', line('.'))
    +    if search('\<from\s\+[A-Za-z0-9._]\{1,50}$', 'bcn', line('.'))
             " Enter character and start completion.
             return "\<space>import \<C-x>\<C-o>\<C-r>=jedi#complete_opened(".a:is_popup_on_dot.")\<CR>"
         endif

**Comment by blueyed** [comment]
Edited the above comment.  Suggested new pattern:

    if search('\<from\s\+[A-Za-z0-9._]\{1,50}\%#\s*$', 'bcn', line('.'))

**Comment by blueyed** [comment]
Also, `is_popup_on_dot` should be `0` always, shouldn't it?

**Comment by blueyed** [comment]
I have tried adding a test:

    it 'smart import'
        exec 'normal ifrom os '
        Expect maparg(" ", "i") == 'jedi#smart_auto_mappings()'
        Expect getline('.') == 'from os import '
    end

But it fails:

    not ok 2 - completions smart import
    # Expected getline('.') == 'from os import '
    #       Actual value: "from os "
    #     Expected value: "from os import "=jedi#complete_opened(0)
    =jedi#complete_opened(0)

I have also tried using `feedkeys` for it.

**Comment by blueyed** [comment]
Regarding the test, it appears that it's the `\%#` (cursor position) that does not get matched.  Doing the `normal` / `call feedkeys()` manually works.. :/

**Comment by davidhalter** [comment]
Yeah testing VIM is a real pain :/

**Comment by aliev** [comment]
@blueyed changed to new pattern in PR

**Comment by blueyed** [comment]
I've managed to get it tested, which is possible when using a `<C-r>` mapping instead:
https://github.com/davidhalter/jedi-vim/compare/smart-import-mapping?expand=1

I'll merge/push this later when CI passes.

**Comment by aliev** [comment]
@blueyed @davidhalter thank you guys! :)

**Comment by aliev** [comment]
@blueyed oh sorry, I think I missed.It may be better to add <silent> for inoremap <buffer> <space> <C-R>=jedi#smart_auto_mappings()<CR>?

**Comment by blueyed** [comment]
@aliev 
You're welcome!  Thank you, too! :octocat:

**Pull request closed without merging**